File: packages/dotnet/NaturalSpacing.Core/OrderedTextUpdateSession.cs
namespace NaturalSpacing.Core;

/// <summary>
/// Coordinates complete hypotheses from revision-capable ASR or dictation
/// providers while retaining only the active utterance ID and revision.
/// </summary>
public sealed class OrderedTextUpdateSession(
    FieldPolicy policy = FieldPolicy.NaturalLanguage,
    OrderedTextSource source = OrderedTextSource.Asr)
{
    private string? _activeUtteranceId;
    private long _lastRevision = -1;

    public FieldPolicy Policy { get; } = policy;
    public OrderedTextSource Source { get; } = source;

    public bool Start(string utteranceId)
    {
        ArgumentNullException.ThrowIfNull(utteranceId);
        if (utteranceId.Length == 0) return false;
        _activeUtteranceId = utteranceId;
        _lastRevision = -1;
        return true;
    }

    public OrderedTextUpdateResult Accept(OrderedTextUpdateEvent update)
    {
        ArgumentNullException.ThrowIfNull(update);
        if (update.Revision < 0) return Rejected(OrderedTextUpdateReason.InvalidRevision);
        if (_activeUtteranceId != update.UtteranceId)
            return Rejected(OrderedTextUpdateReason.InactiveUtterance);
        if (update.Revision <= _lastRevision)
            return Rejected(OrderedTextUpdateReason.StaleRevision);

        _lastRevision = update.Revision;
        var output = NaturalSpacingPolicy.Format(new TextUpdate(
            update.Text,
            Policy,
            Source == OrderedTextSource.Asr ? TextSource.Asr : TextSource.Dictation,
            update.Stability));
        if (update.Stability == TextStability.Final)
        {
            _activeUtteranceId = null;
            _lastRevision = -1;
        }
        return new OrderedTextUpdateResult(true, OrderedTextUpdateReason.Accepted, output);
    }

    public bool Cancel(string utteranceId)
    {
        ArgumentNullException.ThrowIfNull(utteranceId);
        if (_activeUtteranceId != utteranceId) return false;
        _activeUtteranceId = null;
        _lastRevision = -1;
        return true;
    }

    private static OrderedTextUpdateResult Rejected(OrderedTextUpdateReason reason) =>
        new(false, reason, null);
}

File: packages/typescript/core/src/ordered-text-update-session.ts
import { formatTextUpdate } from "./text-update.js";
import type {
  FieldPolicy,
  OrderedTextSource,
  OrderedTextUpdateEvent,
  OrderedTextUpdateResult,
} from "./types.js";

/**
 * Coordinates complete hypotheses from revision-capable ASR or dictation
 * providers. The session retains only the active utterance ID and revision,
 * never hypothesis text.
 */
export class OrderedTextUpdateSession {
  readonly policy: FieldPolicy;
  readonly source: OrderedTextSource;
  #active: { utteranceId: string; lastRevision: number } | null = null;

  constructor({
    policy = "naturalLanguage",
    source = "asr",
  }: {
    policy?: FieldPolicy;
    source?: OrderedTextSource;
  } = {}) {
    this.policy = policy;
    this.source = source;
  }

  start(utteranceId: string): boolean {
    if (utteranceId.length === 0) return false;
    this.#active = { utteranceId, lastRevision: -1 };
    return true;
  }

  accept(event: OrderedTextUpdateEvent): OrderedTextUpdateResult {
    if (!Number.isSafeInteger(event.revision) || event.revision < 0) {
      return { accepted: false, reason: "invalidRevision", output: null };
    }
    if (this.#active?.utteranceId !== event.utteranceId) {
      return { accepted: false, reason: "inactiveUtterance", output: null };
    }
    if (event.revision <= this.#active.lastRevision) {
      return { accepted: false, reason: "staleRevision", output: null };
    }

    this.#active.lastRevision = event.revision;
    const output = formatTextUpdate({
      text: event.text,
      policy: this.policy,
      source: this.source,
      stability: event.stability,
    });
    if (event.stability === "final") this.#active = null;
    return { accepted: true, reason: "accepted", output };
  }

  cancel(utteranceId: string): boolean {
    if (this.#active?.utteranceId !== utteranceId) return false;
    this.#active = null;
    return true;
  }
}
